community/index/src/main/java/org/neo4j/index/internal/gbptree/GroupingRecoveryCleanupWorkCollector.java

```diff
@@ -30,29 +34,44 @@
  */
 public class GroupingRecoveryCleanupWorkCollector implements RecoveryCleanupWorkCollector
 {
-    private final Queue<CleanupJob> jobs = new LinkedList<>();
+    private final Queue<CleanupJob> jobs;
+    private final JobScheduler jobScheduler;
 
-    /**
-     * Separate add phase from run phase
-     */
-    private volatile boolean started;
+    public GroupingRecoveryCleanupWorkCollector( JobScheduler jobScheduler )
+    {
+        this.jobScheduler = jobScheduler;
+        this.jobs = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public void init() throws Throwable
+    {
+        jobs.clear();
+    }
 
     @Override
-    public synchronized void add( CleanupJob job )
+    public void add( CleanupJob job )
     {
-        assert !started : "Tried to add cleanup job after started";
         jobs.add( job );
     }
 
     @Override
-    public synchronized void run()
+    public void start() throws Throwable
     {
-        assert !started : "Tried to start cleanup job more than once";
-        started = true;
         CleanupJob job;
         while ( (job = jobs.poll()) != null )
         {
-            job.run();
+            jobScheduler.schedule( recoveryCleanup, job );
```

Comment:
I may be wrong, but it looks like we're scheduling all cleanup jobs individually. This also means that cleanup jobs could potentially run in parallel. What happens if there are multiple cleanup jobs, each using all available processors? There will be lots and lots of context switching in vain, won't there?